interpreter/interpreter.go

```diff
@@ -38,22 +38,96 @@ type Interpreter interface {
 		decorators ...InterpretableDecorator) (Interpretable, error)
 }
 
+// EvalObserver is a functional interface that accepts an expression id and an observed value.
```

Comment:
Update the docs to indicate the relationship between the `programStep` and `value`. Also document what kinds of values are accepted for the `programStep` value (`Interpretable` and `Qualifier` values)